File: example1/src/main/java/com/example/grpc/Servers.java
package com.example.grpc;

import io.grpc.BindableService;
import io.grpc.Server;
import io.grpc.ServerBuilder;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

public final class Servers {

    public static void start(BindableService bindableService, Logger logger) throws IOException, InterruptedException {
        Loggers.init();

        int port = 50051;
        Server server = ServerBuilder.forPort(port)
            .addService(bindableService)
            .build()
            .start();

        logger.log(Level.INFO, "server started, listening on {0}", port);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            System.err.println("server is shutting down");
            try {
                server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                System.err.println("server shutdown was interrupted");
                server.shutdownNow();
            }
            System.err.println("server has been shut down");
        }));

        server.awaitTermination();
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/Servers.kt
package com.example.grpc

import io.grpc.BindableService
import io.grpc.Server
import io.grpc.ServerBuilder
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

object Servers {
  private val logger = Logger.getLogger(Servers::class.java.name)

  fun start(bindableService: BindableService) {
    Loggers.init()

    val server = ServerBuilder
      .forPort(50051)
      .addService(bindableService)
      .build()
      .start()

    logger.info("server started, listening on ${server.port}")

    Runtime.getRuntime().addShutdownHook(
      Thread {
        System.err.println("server is shutting down")
        try {
          server.shutdown().awaitTermination(10, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
          server.shutdownNow()
        }
        System.err.println("server has been shut down")
      }
    )

    server.awaitTermination()
  }

  fun start(bindableService: BindableService) {
    val serverBuilder = ServerBuilder
      .forPort(50051)
      .addService(bindableService)

    start(serverBuilder)
  }
}
